Sources/Networking/Core/Upload/UploadAPIManager.swift

```diff
@@ -0,0 +1,286 @@
+        guard var existingUploadTask = await uploadTasks.getValue(for: request.id) else {
+            return UploadTask(
+                sessionUploadTask: sessionUploadTask,
+                endpointRequest: request,
+                uploadable: uploadable
+            )
+        }
+        existingUploadTask.task = sessionUploadTask
+        return existingUploadTask
+    }
+
+    func handleUploadTaskCompletion(
+        urlRequest: URLRequest,
+        endpointRequest: EndpointRequest,
+        retryConfiguration: RetryConfiguration?,
+        data: Data?,
+        response: URLResponse?,
+        error: Error?
+    ) {
+        Task {
+            guard let uploadTask = await uploadTasks.getValue(for: endpointRequest.id) else {
+                return
+            }
+
+            var state = UploadTask.State(task: uploadTask.task)
+            if let data, let response {
+                state.response = try await responseProcessors.process(
+                    (data, response),
+                    with: urlRequest,
+                    for: endpointRequest
+                )
+
+                try await uploadTask.complete(with: state)
+
+                // Cleanup on successful task completion
+                await uploadTask.resetRetryCounter()
+                await uploadTasks.set(value: nil, for: endpointRequest.id)
+            } else if let error {
+                do {
+                    // URLError.Code.cancelled is thrown if the URLSessionTask is cancelled.
+                    // Consider this action intentional, thus the request won't be retried.
+                    guard !state.cancelled else {
+                        throw error
+                    }
+
+                    try await uploadTask.sleepIfRetry(
+                        for: error,
+                        retryConfiguration: retryConfiguration
+                    )
+
+                    try await self.uploadRequest(
+                        uploadTask.uploadable,
+                        request: uploadTask.endpointRequest,
+                        retryConfiguration: retryConfiguration
+                    )
+                } catch {
+                    state.error = await errorProcessors.process(
+                        error,
+                        for: uploadTask.endpointRequest
+                    )
+
+                    // No cleanup in case the task will be retried.
+                    try await uploadTask.complete(with: state)
+                }
+            }
+        }
+    }
+
+    func sessionUploadTask(
+        with uploadable: Uploadable,
+        for request: URLRequest,
+        completionHandler: @escaping @Sendable (Data?, URLResponse?, Error?) -> Void
+    ) -> URLSessionUploadTask {
+        switch uploadable {
+        case let .data(data):
+            return urlSession.uploadTask(
+                with: request,
+                from: data,
+                completionHandler: completionHandler
```

Comment:
why not use the async version?

Reply:
Good question! 🤔 We could definitely consider using the async version, however, if we go down that route, we'll need to make some slight changes to the implementation, because:
- The async versions don't return `URLSessionTask` and will be suspended, which means returning immediately as it is now wouldn't be possible. So we'll need to consider what to return to the client, if anything at all.
- We'll need to handle errors and responses from the request using delegates.

If you have any ideas or suggestions, please feel free to share them. 🙌

Reply:
Ah, that's unfortunate but I don't see a way to make it work as conveniently with the async version, as you're saying. Let's just keep it as is :)

Reply:
We can explore an alternative approach by wrapping the asynchronous part of the request in a `Task`, which can be returned in the `UploadTask` object (this would give us some control over the lifetime of that Task too). However, as you mentioned, this approach may not be as convenient as using the completion closure.

Reply:
Agree, I would keep it as it is. Even though async/await is cool, still older approaches (closures, combine) still their own purposes and I don't mind mixing them if it makes sense. I would just consider to put @tonyskansf explanation into code as a comment. Best place may be the protocol where this method is defined.